pipeline_dp/dp_engine.py

```diff
@@ -49,3 +49,57 @@ def aggregate(self, col, params: AggregateParams,
     # TODO: implement aggregate().
     # It returns input for now, just to ensure that the an example works.
     return col
+
+  def _bound_contributions(self, col,
+                           max_partitions_contributed: int,
+                           max_contributions_per_partition: int,
+                           aggregator_fn):
+    """
+    Bounds the contribution by privacy_id in and cross partitions.
+    Args:
+      col: collection, with types of each element: (privacy_id,
+        partition_key, value).
+      max_partitions_contributed: maximum number of partitions that one
+        privacy id can contribute to.
+      max_contributions_per_partition: maximum number of records that one
+        privacy id can contribute to one partition.
+      aggregator_fn: function that takes a list of values and returns an
+        aggregator object which handles all aggregation logic.
+
+    return: collection with elements ((privacy_id, partition_key),
+          aggregator).
+    """
+    # per partition-contribution bounding with bounding of each contribution
+    col = self._ops.map_tuple(col, lambda pid, pk, v: ((pid, pk), v),
+                              "Rekey to ( (privacy_id, partition_key), value))")
+    col = self._ops.sample_fixed_per_key(col,
+                                         max_contributions_per_partition,
+                                         "Sample per (privacy_id, partition_key)")
+    # ((privacy_id, partition_key), [value])
+    col = self._ops.map_values(col, aggregator_fn,
+                               "Apply aggregate_fn after per partition bounding")
+    # ((privacy_id, partition_key), aggregator)
+
+    # Cross partition bounding
+    col = self._ops.map_tuple(col, lambda pid_pk, v: (pid_pk[0],
+                                                      (pid_pk[1], v)),
+                              "Rekey to (privacy_id, (partition_key, "
+                              "aggregator))")
+    col = self._ops.sample_fixed_per_key(col, max_partitions_contributed,
+                                         "Sample per privacy_id")
+    # (privacy_id, [(partition_key, aggregator)])
+    return self._ops.flat_map(col,
+                              self._unnest_cross_partition_bound_sampled_per_key,
+                              "Unnest")
+
+  def _unnest_cross_partition_bound_sampled_per_key(self, pid):
```

Comment:
please rename pid to pid_pk_v (since it's not pid, but a tuple)

Reply:
Done